source/agora/node/FullNode.d

```diff
@@ -360,21 +360,23 @@ public class FullNode : API
             preimages = the preimages needed to check the validity of the blocks
 
         Returns:
-            true if the blocks and preimages was added
+            the last read block height
 
     ***************************************************************************/
 
-    public bool addBlocks (const(Block)[] blocks, const(PreImageInfo)[] preimages)
+    public Height addBlocks (const(Block)[] blocks, const(PreImageInfo)[] preimages)
         @safe
     {
+        Height last_height;
         foreach (block; blocks)
         {
             if(!this.ledger.enrollment_manager.addPreimages(preimages))
```

Comment:
This line also looks super buggy, especially now with preimages being potentially disjoint between cycles.
If we get blocks [18, 19, 20, 21] and pre-images for 21, the EM would reject them ATM.

Reply:
Oh then I think we should add a `rollback()` call here.

Reply:
Now that I look at this function more it looks really sketchy. `addPreimages` is called multiple times...

Reply:
I think we need to rework the catching up phase w.r.t. preimages. I'll file a separate issue.

Reply:
Scratch that, I'll just fix the function. I've also noticed that this function will return `false` if we receive a duplicate preimage (same height), which means that if we receive a newer preimage and *then* do a `getBlocksFrom + getPreimages` the `addBlocks` function will terminate early and the blocks would be rejected.

Reply:
IMO it's still broken ATM, because upon receiving PI for 21, the Ledger (at 17) will reject them and so adding block 18 will not work. When we move from distance to height, we will need to accept PI beyond the enrolled cycle (or adjust them to be at the end of the cycle, but there's no downside in accepting them beyond the cycle within a reasonable margin).